### PR TITLE
In collection, Add an enforced state to the modules

### PR DESCRIPTION
**Pull request by sean-m-sullivan**
##### SUMMARY
Follow up PR from #13927, this is expected to fail tests until #13890 is merged, as it contains code from that PR

This enforces default values being submitted to the API, this PR is an overview of how that enforcement mechanism would be achieved across the collection, This PR is for feedback on how this was achieved so that I can go through and add it to the rest of the modules. The idea is to enforce a state to prevent drift when an option is not provided and option is enforced. This is compared to exists that makes no changes if the object exists, or present that only makes changes for the options given to the module.

##### ISSUE TYPE
 - New or Enhanced Feature

##### COMPONENT NAME
 - Collection

##### AWX VERSION

```
22.1.0
```

### CODE REVIEW

**Comment by AlanCoding** [comment]
@john-westcott-iv brought up a good proposal about using OPTIONS to automatically get the defaults.

**Comment by AlanCoding** [comment]
I have a question - what about related lists? I want to have that behave coherently given the state options of exists or enforced.

Logically, the "default" value for a related list (like credentials for a job template) would be an empty list. However, users have the ability to not specify credentials, in which case, they will not be adjusted in any way. If someone wanted it "enforced" and the JT _has_ related credentials already, should they be removed? I'm leaning towards "yes".

What about the "exists" option? If creating a new JT, should it associate related credentials? Strong lean towards "yes". Should the normal "present" state enforce the list is _exactly equal_ to the provided list, or just that the items specified are there? I lean towards the former (current behavior), but worry this might latter get us in some kind of catch 22 I haven't thought about.

**Comment by sean-m-sullivan** [comment]
@AlanCoding This is why it is piggybacked on the state, 
If state == present, things behave EXACTLY as they have in the past, if State == exists, it exits out after confirming existance, ONLY if state == enforced does it start overwriting things, its why you were right in adding this in as a state option, as they cannot coexist.

**Comment by AlanCoding** [comment]
My near-term question here is - if I use the `job_template` module with state=enforced, and it currently has credentials, and the module does not specify any credentials, does it remove existing credentials?

I think we can work this out quickly and efficiently here. I would say, yes, it _should_ remove credentials and any objects in any other relationships like this.

File: awx_collection/plugins/modules/schedule.py [hunk 1]
```diff
@@ -265,15 +265,29 @@ def main():
...
+    # Fail early now if no unified job template provided now that enforced defaults set.
+    if unified_job_template is None and enforced_defaults:
+        module.fail_json(msg='When using enforced state unified_job_template is a required field')
+
     if credentials is not None:
         association_fields['credentials'] = []
         for item in credentials:
             association_fields['credentials'].append(module.resolve_name_to_id('credentials', item))
+    elif credentials is None and enforced_defaults:
+        association_fields['credentials'] = []
```
Comment (sean-m-sullivan) [inline awx_collection/plugins/modules/schedule.py:290]: @AlanCoding This line I tested removes existing credentials when submited to the API. when none are supplied and enforced is true.

Comment (sean-m-sullivan) [reply awx_collection/plugins/modules/schedule.py:290]: This would be the same as in a JT

**Pull request closed without merging**